### ISSUE TITLE
Cache credentials and add support for MFA usage with STS assume role

### ISSUE THREAD

**Issue by jonknapp** [serverless/serverless#1110]
I wanted to start a conversation around how we could support multi-factor authentication when using STS `assumeRole` for AWS calls. I have put together a simple example to show how it could integrate with the current codebase, but it has a few downsides. The code is in a [pull request](https://github.com/CoffeeAndCode/serverless/pull/1) on my fork that I will gladly bring over if desired.
- I am caching AWS credentials as multiple calls with the same MFA token will fail.
- I need to pass the MFA token in an environment variable since the application does not have the code to pull from stdin if a code is required.
- There are a lot of calls to `getRoleCredentials` that are made for deploying a single function (as there are many parts to a simple function deploy). I'm not sure how caching credentials negatively impacts the codebase.

In a perfect world, we would request the MFA on stdin when we see that the aws profile is using a role with MFA or an access request fails with an invalid MFA code, but I'm sure someone here would be able to help that happen faster than I could right now.

There are no tests in this pull request at this point.

To use, you would make your normal `serverless` cli calls but have the environment variable `AWS_MFA_TOKEN` set to the current timed code. Since the credentials are cached, we only need it to be the current code when the first call is made.

`AWS_MFA_TOKEN=123456 serverless dash deploy`

**Comment by jonknapp** [comment]
I should mention even though my code caches credentials, if I attempt to deploy two functions at the same time it fails. I didn't try to look into why at this point.

**Comment by jonknapp** [comment]
.... morning, no coffee. accidental click.  sorry.

**Comment by jonknapp** [comment]
I solved this for myself by writing a small application in charge of getting AWS temp credentials from an STS `assumeRole` call, then outputting them as bash environment variables to `stdout`.  Then I can eval it to have other applications take advantage of the new environment variables.

The app is [aws-auth-helper](https://www.npmjs.com/package/aws-auth-helper) on npm.

Here's a snippet from my `package.json` file:

```
"scripts": {
    "deploy": "eval \"`aws-auth-helper` serverless dash deploy\""
}
```

which allows me to run `npm run deploy` in bash on my mac to allow two-factor auth deploys. It basically ends up calling a command that looks like:

```
AWS_ACCESS_KEY_ID=xxx AWS_SECRET_ACCESS_KEY=xxx AWS_SESSION_TOKEN=xxx serverless dash deploy
```

This functionality works for me, but it's pretty limited in scope (requires bash). A longer term solution for serverless would probably involve new authentication code or allowing authentication to be handled by a plugin (which it may already be able to do...).

**Comment by flomotlik** [comment]
@jonknapp are you using this inside of the functions? I'm going to close it for now as I don't think we're going to support this anytime soon (haven't heard a lot about it really). But looks interesting

**Comment by jonknapp** [comment]
I'm using `aws-auth-helper` as a command that I call before the `serverless dash deploy` command so that aws environment variables can be exported and available when serverless runs.

Ideally, calling the `aws.assumeRole` method would be done by a Serverless plugin so it wouldn't matter what shell the user was using, but if there's not enough demand for it I'll just stick with my bash helper script.  Thanks for looking it over.

**Comment by flomotlik** [comment]
@jonknapp thanks for the details. I'll keep it in mind when I hear more 2FA support request.

**Comment by martinmicunda** [comment]
+1 for this feature.. I am using cross account access and MFA is requirement for me.

**Comment by jonknapp** [comment]
@martinmicunda If you check out [aws-auth-helper](https://www.npmjs.com/package/aws-auth-helper), let me know what you think. You can ping me at jon@coffeeandcode.com.

**Comment by jthomerson** [comment]
@flomotlik could we re-open this? Needing support for MFA is a very valid usecase. If you're using cross-account credentials especially this comes in handy.

**Comment by flomotlik** [comment]
Sounds good @jthomerson , reopened. Though we won't be able to dedicate resources ourselves to it at this time, but we'd love to get somebody to work on this.

Btw we're using the same setup with MFA for all of our accounts and have had a good run with using the assume-role tool: https://github.com/remind101/assume-role

**Comment by jthomerson** [comment]
Ah, thanks for the tip @flomotlik ... I'm going to try the assume-role tool. If that works well then maybe SLS doesn't really need to add support for MFA to its scripts, but instead just have documentation on how to use this tool. I mean, this is a framework designed to support microservices, so it seems in line with that principle to use a different tool rather than building SLS into a monolith that handles every case. 😄

**Comment by flomotlik** [comment]
@jthomerson I'd still like to have MFA support, but its just not enough of a pain at the moment that we want to put resources on it. If somebody wants to do it though more power to them :).

**Comment by kiwifellows** [comment]
Hi,
Just noticed this after looking for this feature myself.

Some ideas I have are to have the Cross Account IAM roles feature work along the lines of:
configure a new item in serverless.yml  under provider "aws" called "deployroles" or similar.

Also in terms of standardising the approach as well as ensuring features are only available depending on the provider. E.g AWS

So in the YAML this could look like:
```yaml
# Serverless config for some serverless stuff
service: myServerlessMicroService

provider:
  name: aws
  runtime: nodejs4.3
  region: ap-southeast-2
  deployroles:
     mydev: arn:aws:iam::...:role/MyIAMRoleName-Dev-27PFE...
     mytest: arn:aws:iam::...:role/MyIAMRoleName-Test-27PFE...
     myprod: arn:aws:iam::...:role/MyIAMRoleName-Prod-27PFE...
  memorySize: 512
  iamRoleStatements....
```

As a command it could potentially look like this:
```bash
serverless deploy --deployrole mytest
```

Obviously this would be quite a change to the underlying YAML, but could make sense.

In terms of the implementation of this in serverless I'm unsure exactly how would go about modifying this in the current code base so I could implement STS / Assume Role. If someone gave me a starting point I wouldn't mind working towards a PR.

cheers,

Ben

**Comment by eric-tucker** [comment]
From a security standpoint, this should be prioritized much higher.  MFA is an AWS best practice, especially when using cross account which is also a recommended best practice as AWS recommends one account for IAM users (named a security account) and all other resources go into separate accounts.

By not supporting this you are telling your users that other features are more important than security.

**Comment by kiwifellows** [comment]
Hi @eric-tucker some thoughts on your recommended approach.

MFA is a best practice for AWS IAM user accounts that need to access their own AWS account and do Cross Account Role access to other AWS accounts. 

So I'm assuming you're talking about using MFA for when using serverless deploy on a dev machine?

In terms of production, it is best to use roles that do cross account access from one IAM role in AWS Account A wanting to STS Assume Role to Account B.

The IAM role used in Account A to Cross account/STS assume to Account B should be using the proper security context and with the right IAM policies to limit access to only the resources required. I don't think supporting MFA is a "must" for serverless to be secure as Access Keys/ Secrets for dev accounts should have some rotation policy if the company deems it as a risk mitigation.

What should be a security practice is for customers using serverless and AWS to get away from using AWS access keys/secrets in production environments/deployments. Ideally deployments are done in a production environment from an EC2 instance that has serverless installed and the EC2 instance uses an appropriate IAM Instance profile. 

Again it's for individual customers to figure out how they secure their AWS environment with the right expertise.

Just some thoughts around Dev vs. Prod deployments... feel free to feedback :)

**Comment by eric-tucker** [comment]
All completely valid points, but I still believe that supporting MFA should be a priority.

We can work around this for now by writing a custom shell script which does the assume role call then exports the standard variables:
* AWS_ACCESS_KEY_ID
* AWS_SECRET_ACCESS_KEY
* AWS_SESSION_TOKEN

So this isn't a blocker, but a very nice to have.

**Comment by eric-tucker** [comment]
I'd like to come back to this as I didn't quite catch this statement the first time:

>  I don't think supporting MFA is a "must" for serverless to be secure as Access Keys/ Secrets for dev accounts should have some rotation policy if the company deems it as a risk mitigation.

This is questionable advice.  You're telling people that rotating keys is a sufficient solution in place of using MFA.  That is a fundamentally flawed stance to take.

**Comment by str3tch** [comment]
I'd like to add support for MFA within Serverless. I work in a financial institution, and MFA on API use makes things go much better with security!

**Comment by kiwifellows** [comment]
HI @et304383 
All I'm saying is I don't agree MFA should be as high as you're wanting it to be. Most automated deployments I see across different production and non-production environments are done without MFA, the main reason being that it doesn't involve a human.

I've provided independent consulting to several different parties using AWS and always in partnership with the AWS team the advice has been "use the right security framework and approach for the outcome you're wanting.".  Sometimes because it's security related we believe we need to use the same hammer to achieve the same outcome.

If the problem you're trying to solve is enforcing MFA for developers when doing a "serverless deploy" I'd almost argue that access key and secret is enough. I'd also argue that dev and production AWS accounts should be completely separate and that if you have unauthorised access to a DEV AWS account it can be mitigated failry easily using in built AWS tools (Cloud Trail etc) and 3rd party tools. I'd also argue that once you get to deploying production serverless environments, these would be done using tooling that sits inside a VPC (if your individual security requirements impose that on you) and done using automated deployments. One of the problems I see when people bring a traditional view of security into the public cloud arena, they already forget AWS has already provided a massive investment into security Day 1. So to say that Access Keys/Secrets with a rotation policy based on business need aren't secure would be a false and misleading statement. This is what AWS would agree with (that key rotation is a good security practice). 

I've used MFA a lot in both prod and dev environments on AWS, but typically I am using MFA to limit and reduce risk of unauthorised access on AWS Console or using the AWS CLI for adhoc commands. Generally in a dev AWS environment where we are using native AWS tools (Lambda, Dynamo, SNS, S3 etc...) I am trusting developers with a bit of opening up these services, still within the context of a Dev environment.

P.S @str3tch I completely understand your viewpoint of having better security for a financial instution (and in comparison most organisations that handle either personal, health or payment data should have a very minimal set of controls in place), but I just think that MFA is not always the right hammer when you're looking at deployment processes (rather than just ad hoc AWS access). I hope that makes sense.

**Comment by jthomerson** [comment]
@str3tch and @et304383 please see https://github.com/remind101/assume-role, which @flomotlik linked to above. I've used that for well over a year now in my day-to-day development, and am very happy with that. I actually prefer this now over having MFA built straight into Serverless. Why? Because I can do `eval $(assume-role DEV)` and that shell will continue to allow me to run Serverless commands for an hour. If MFA were built right into Serverless, I'd have to re-enter my MFA token for every command ... and can't run two commands in quick succession, because I'd have to wait on a new token value. I'd suggest giving it a shot and seeing if that meets your needs as well.

@kiwifellows be careful about giving security advice that goes against documented best practices. Using a key and secret is never enough for any environment you care dearly about (including dev), because they are long-term credentials. Sure, your developer will hopefully treat them right, but what if your help desk guy copies the `~/.aws` directory while he's working on your developer's machine, and now he has the key and secret? That developer stays around, and doesn't rotate his keys for X months ... your help desk guy can use them until they're invalidated. If your developers used MFA, your help desk guy couldn't do anything with those credentials. In this scenario I'm using a help desk guy - even though they're generally 'on your side' - just because they have easy access to your workstations. But there are many bad actors that are definitely not on your side and could easily get credentials.

**Comment by kiwifellows** [comment]
Hi @jthomerson,
First, thanks for the honest feedback and I agree in principle what you're saying.

Respectfully, I understand what you're saying about documented best practices. AWS would not design Access/Secrets as a method of connecting securely to AWS APIs if they thought there was no mitigation. As I said earlier key rotation is one approach, but if you can't enforce key rotation because you've given devs more open access then I'm suggesting other approaches are to use a mix of AWS provided tools such as AWS Cloud Trail and 3rd party tools that can help mitigate the risk of someone unauthorized doing something to an AWS account and immediately highlight to a sys admin team any out of normal behaviour (defined by the team). I did forget to mention of course that there are a number of different other approaches that could be used... e.g. look at how IAM policies are put together etc. etc...

If a sys admin is really not going to trust that his access/secrets aren't going to work then as mentioned there are the options of STS (and we can revoke/expire access for any STS tokens), MFA as suggested and even look at using EC2 machines for deployment that have IAM instance profiles...

Again each use case is different and depending on the deployment scenarios for developers, testers, automated:yes/no, prod etc....

Good discussion to have anyway, so thanks.

Ben

**Comment by chizou** [comment]
If MFA is implemented, it should be implemented with a cache (like aws-cli's implementation) so as to avoid needing to add MFA credentials every time as @jthomerson is mentioning.

That said, there are a lot of inaccuracies being stated in this thread. 

First off, Key rotation is **never** enough. If you have a careless developer that commits their credentials and posts it to a public Github repo, there are numerous bots scanning Github and will exploit it within minutes to hours. No key rotation policy can possibly protect you against that. Implementing any of the other solutions you mentioned (e.g. autoamted CloudTrail detection)  requires significant overhead that the huge majority of AWS users don't have resources to implement. Conversely, MFA is trivial for every user to enable.

Secondly, automated deployment systems shouldn't be using access/secret keys either. They should be using IAM instance profiles / IAM roles, as stated in the AWS Well Architected Framework. So saying MFA isn't needed because automated systems don't need them is a false equivalency. 

Just because use cases are different doesn't mean they're good or even acceptable. Ultimately, it's not a good reason to not implement native MFA support.

**Comment by alikian** [comment]
I created an MFA plug-in here https://github.com/alikian/mfa-serverless-plugin

**Comment by jplock** [comment]
Is there any interest in incorporating @alikian's plugin into the main codebase?  I've found it very useful

**Comment by jplock** [comment]
@alexdebrie any thoughts on @alikian's plugin https://github.com/alikian/mfa-serverless-plugin and if something like MFA support could be supported natively within Serverless?

**Comment by KaloyanP** [comment]
Is anyone looking into resolving this issue? Not having workable MFA i.e. having to type in an MFA token for every deployment, makes the serverless framework very difficult to use professionally without a third party auth tool. MFA is pretty much a standard security requirement for large enterprises, finance, and government - having proper support for MFA should be a no-brainer.

While we can mitigate against this externally, the fact that authentication behaviour between the AWS CLI and Serverless is different makes me *very* anxious. Why is there a difference, when the code already exists to cache my credentials? What part of the the AWS CLI authentication might have been reimplemented insecurely?

**Comment by alikian** [comment]
@KaloyanP I agree serverless need natively support MFA, because I don't know at when an change in code may break this plugin. I can make a PR for that but not sure will get approval.

**Comment by tijoer** [comment]
This is kinda critical if you are working with MFA. alikian solution could be pulled into the main sls code.

**Comment by medikoo** [comment]
MFA support was kind of implemented with #5432, still it remains broken at this point due to: https://github.com/serverless/serverless/issues/6576

I'm going to close this, follow #6576 for updates on MFA support fix
